File: rust/advent-of-code-2021/src/main.rs
use solutions::{
    eight,
    eleven::{read_octopus_from_file, simulate_flash, sync_flash},
    five, four, nine, one, seven, six, ten, three, two,
};

mod solutions;

fn main() {
    // exec_one();
    // exec_two();
    // exec_three();
    // exec_four();
    // exec_five();
    // exec_six();
    // exec_seven();
    // exec_eight();
    // exec_nine();
    // exec_ten();
    exec_eleven();
}

#[allow(dead_code)]
fn exec_one() {
    let sweeps = one::read_sweep_from_file();

    let increase_sweep_count = one::count_increase_sweep(&sweeps);
    println!("The number of increase sweep is {}", increase_sweep_count);

    let increase_sweep_window_count = one::count_increase_sweep_window(&sweeps);
    println!(
        "The number of increase sweep window is {}",
        increase_sweep_window_count
    );
}

#[allow(dead_code)]
fn exec_two() {
    let commands = two::read_commands_from_file();

    let (horizontal, depth) = two::simulate_commands(&commands);
    println!(
        "Multiple of horizontal position and depth is {}",
        horizontal * depth
    );

    let (final_horizontal, final_depth) = two::simulate_commands_with_aim(&commands);
    println!(
        "With aim: Multiple of horizontal position and depth is {}",
        final_horizontal * final_depth
    );
}

#[allow(dead_code)]
fn exec_three() {
    let bits = three::read_bits_from_file();

    let power_consumption = three::process_bits(&bits);
    println!("The power consumption is {}", power_consumption);

    let oxygen_rating = three::find_oxygen_rating(&bits);
    let co2_rating = three::find_co2_rating(&bits);
    println!("{}", oxygen_rating * co2_rating);
}

#[allow(dead_code)]
fn exec_four() {
    let (numbers, mut bingos) = four::read_bingo_from_file();

    let (step, point) = four::calculate_fastest_win(&numbers, &mut bingos);
    println!("Fastest win with {} step and {} points", step, point);

    let (step, point) = four::calculate_slowest_win(&numbers, &mut bingos);
    println!("Slowest win with {} step and {} points", step, point);
}

#[allow(dead_code)]
fn exec_five() {
    let vent_lines = five::read_vents_from_file();
    let intersections = five::count_intersection(&vent_lines);
    println!("Number of line intersections: {}", intersections);

    let intersections_with_diagonal = five::count_intersection_with_diagonal(&vent_lines);
    println!(
        "Number of line intersections, include diagonal line: {}",
        intersections_with_diagonal
    );
}

#[allow(dead_code)]
fn exec_six() {
    let remaining_days = 256;
    let intervals = six::read_interval_from_file();

    let lanternfish_count = six::count_lanternfish_optimize(&intervals, remaining_days);

    println!("#fish after {} days: {}", remaining_days, lanternfish_count);
}

#[allow(dead_code)]
fn exec_seven() {
    let positions = seven::read_position_from_file();

    let min_medium = seven::min_horizontal_medium(&positions);
    println!("{}", min_medium);

    let min_increment_medium = seven::min_horizontal_medium_increment(&positions);
    println!("{}", min_increment_medium);
}

#[allow(dead_code)]
fn exec_eight() {
    let signals = eight::read_signal_from_file();

    let unique_signal_count = eight::count_unique_signals(&signals);
    println!("Number of unique signal {}", unique_signal_count);

    let decode_signal_sum = eight::sum_decode_signal(&signals);
    println!("Decode sum {}", decode_signal_sum);
}

#[allow(dead_code)]
fn exec_nine() {
    let locations = nine::read_location_from_file();

    let risk_sum = nine::sum_risk_level(&locations);
    println!("Risk index {}", risk_sum);

    let mul_basin = nine::mul_top_three_basins(&locations);
    println!("Multiple of top 3 basins {}", mul_basin);
}

#[allow(dead_code)]
fn exec_ten() {
    let lines = ten::read_code_from_file();

    let illegal_point = ten::sum_illegal_point(&lines);
    println!("Illegal point: {}", illegal_point);

    let autocomplete_point = ten::sum_autocomplete_point(&lines);
    println!("Autocomplete point: {}", autocomplete_point);
}

#[allow(dead_code)]
fn exec_eleven() {
    let octopus_matrix = read_octopus_from_file();

    // Clone to avoid mutating each other input
    let flash_count = simulate_flash(&mut octopus_matrix.clone(), 100);
    println!("Flash count after 100 steps: {:?}", flash_count);

    // Clone to avoid mutating each other input
    let sync_step = sync_flash(&mut octopus_matrix.clone());
    println!("Octopus sync at {}", sync_step);
}

File: golang/advent-of-code-2021/main.go
package main

import (
	"aoc21/five"
	"aoc21/four"
	"aoc21/one"
	"aoc21/seven"
	"aoc21/six"
	"aoc21/three"
	"aoc21/two"
	"fmt"
	"math"
)

func main() {
	// execOne()
	// execTwo()
	// execThree()
	// execFour()
	// execFive()
	// execSix()
	execSeven()
}

//lint:ignore U1000 Old solutions
func execOne() {
	sweeps := one.ReadSweepFromFile()

	increase_sweep := one.CountIncreaseSweep(&sweeps)
	fmt.Printf("The number of increase sweep %d\n", increase_sweep)

	increase_sweep_range := one.CountIncreaseSweepRange(&sweeps)
	fmt.Printf("The number of increase sweep range %d\n", increase_sweep_range)
}

//lint:ignore U1000 Old solutions
func execTwo() {
	commands := two.ReadCommandFromFile()

	horizontal, vertical := two.SimulateCommands(&commands)
	fmt.Printf("Multiple of horiontal position and vertical position is %d\n", horizontal*vertical)

	horizontalWithAim, verticalWithAim := two.SimulateCommandsWithAim(&commands)
	fmt.Printf("With aim: Multiple of horiontal position and vertical position is %d\n", horizontalWithAim*verticalWithAim)
}

//lint:ignore U1000 Old solutions
func execThree() {
	bits := three.ReadBitsFromFile()

	powerConsumption := three.CalculatePowerConsumption(bits)
	fmt.Printf("Power consumption %d\n", powerConsumption)

	oxygenRating := three.CalculateOxygenRating(bits)
	co2Rating := three.CalculateCO2Rating(bits)
	fmt.Printf("Life rating %d\n", oxygenRating*co2Rating)
}

//lint:ignore U1000 Old solutions
func execFour() {
	bingoNumbers, bingos := four.ReadBingoFromFile()

	// Part 1
	minWinStep := math.MaxInt
	point := 0
	for _, bingo := range bingos {
		bingoWinStep, bingoPoint := four.CheckBingo(bingoNumbers, bingo)
		if bingoWinStep < minWinStep {
			minWinStep = bingoWinStep
			point = bingoPoint
		}
	}
	fmt.Println("Fatest win at", minWinStep, "with point", point)

	// Part 2
	maxWinStep := 0
	for _, bingo := range bingos {
		bingoWinStep, bingoPoint := four.CheckBingo(bingoNumbers, bingo)
		if bingoWinStep > minWinStep {
			minWinStep = bingoWinStep
			point = bingoPoint
		}
	}
	fmt.Println("Slowest win at", maxWinStep, "with point", point)
}

//lint:ignore U1000 Old solutions
func execFive() {
	ventLines := five.ReadVentFromFile()
	intersections := five.CountIntersections(&ventLines)
	fmt.Printf("Number of intersections %d\n", intersections)

	intersectionDiagonal := five.CountIntersectionsWithDiagonal(&ventLines)
	fmt.Printf("Number of intersections with diagonal %d\n", intersectionDiagonal)
}

//lint:ignore U1000 Old solutions
func execSix() {
	intervals := six.ReadFishIntervalsFromFile()

	days := 2
	count := six.CountFish(&intervals, days)

	fmt.Printf("Fishes after %d day(s): %d\n", days, count)
}

//lint:ignore U1000 Old solutions
func execSeven() {
	crabs, err := seven.ReadCrabFromFile()
	if err != nil {
		panic(err)
	}

	crabMoves := seven.MinCrabMove(crabs)
	println("Minimum crab move", crabMoves)

	crabExtraMoves := seven.MinCrabExtraMove(crabs)
	println("Minimum crab extra move", crabExtraMoves)
}
